src/ogma/features/nodeGrouping.ts

```diff
@@ -0,0 +1,205 @@
+   * @param ruleName The nae of node grouping rule
+   * @param type the type of the node
+   * @param property the property name that will be used to group the nodes
+   */
+  public setGroupingRule(ruleName: string, type: string, property: string): void {
+    this.groupRule = {ruleName: ruleName, type: type, property: property};
+  }
+
+  /**
+   * create a node grouping transformation
+   * It uses groupRule to define the rule
+   * Group the nodes based on a category type and a property value
+   */
+  public async initTransformation(): Promise<void> {
+    if (this.transformation === undefined) {
+      this.transformation = this._ogma.transformations.addNodeGrouping({
+        groupIdFunction: (node) => {
+          if (this._isRuleNotApplicableToNode(node)) {
+            return undefined;
+          } else {
+            const propertyValue = node.getData(['properties', this.groupRule?.property ?? '']);
+            const originalValue =
+              typeof propertyValue === 'object'
+                ? (propertyValue as ConflictValue).original
+                : propertyValue;
+            // groupRule is defined if not we returned undefined
+            return `${this.groupRule!.type}-${originalValue}`.toLowerCase().trim();
+          }
+        },
+        nodeGenerator: (nodes) => {
+          return {
+            data: {
+              categories: [this.groupRule?.type],
+              properties: {
+                size: nodes.size
+              }
+            }
+          };
+        },
+        edgeGenerator: () => {
+          return {
+            data: {
+              type: LKE_NODE_GROUPING_EDGE
+            }
+          };
+        },
+        showContents: true,
+        duration: 300,
+        padding: 10
+      });
+      this._listenToTransformationEvents();
+    } else {
+      await this.refreshTransformation();
+    }
+  }
+
+  /**
+   * refresh the transformation
+   * Called when there is a change in the rule
+   */
+  public async refreshTransformation(): Promise<void> {
+    await this.transformation?.refresh();
+  }
+
+  /**
+   * init node grouping style
+   */
+  public initNodeGroupingStyle(): void {
+    this.nodeGroupingStyleRule = this._ogma.styles.addRule({
+      nodeAttributes: {
+        // Any default style will go here
+        text: {
+          content: (node: Node<LkNodeData> | undefined): string | undefined => {
+            return this._getNodeGroupingCaption(node);
+          },
+          style: 'bold'
+        },
+        layer: -1,
+        opacity: 0.32
+      },
+      nodeSelector: (node) => {
+        return node.isVirtual();
+      },
+      nodeDependencies: {self: {data: true}}
+    });
+  }
+
+  /**
+   * run layout on all subnodes of virtual nodes
+   */
+  public async runLayoutOnAllSubeNodes(): Promise<void> {
+    // @ts-ignore getContext exists on the transformation but hidden by the types
+    const virtualNodes = this.transformation.getContext().virtualNodes;
```

Comment:
just to be sure: did you check with the ogma team this is the correct way to do it? Why is it hidden?

Reply:
yes I did, it was their proposition. Why, for internal decision..